### PR TITLE
Fix #216: Consent for Onboarding

### LINKED ISSUE #216: Consent for Onboarding
Implement consent for onboarding.

- obtain consent text
- store approval

### PR DESCRIPTION
**Pull request by banterCZ**

### CODE REVIEW

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/controller/api/OnboardingController.java [hunk 1]
```diff
@@ -183,4 +187,50 @@ public Response performCleanup(@EncryptedRequestBody ObjectRequest<OnboardingCle
...
+            description = "Obtain a text of user consent in specified language."
+    )
+    public ObjectResponse<OnboardingConsentTextResponse> fetchConsentText(final @RequestBody OnboardingConsentTextRequest request) throws OnboardingProcessException{
+        logger.debug("Returning consent for {}", request);
+        validate(request);
+        final OnboardingConsentTextResponse onboardingConsentTextResponse = onboardingService.fetchConsentText(request);
+        return new ObjectResponse<>(onboardingConsentTextResponse);
+    }
+
+    @PostMapping("consent/approve")
```
Comment (banterCZ) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/controller/api/OnboardingController.java:202]: I am open to discuss this path.

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/OnboardingService.java [hunk 2]
```diff
@@ -316,6 +318,58 @@ public void terminateInactiveProcesses() {
...
+                .build();
+
+        try {
+            onboardingProvider.approveConsent(providerRequest);
+        } catch (OnboardingProviderException e) {
+            throw new OnboardingProcessException("An error when approving consent.", e);
+        }
+    }
+
+    private static com.wultra.app.onboardingserver.provider.ConsentType convert(final ConsentType source) {
```
Comment (banterCZ) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/OnboardingService.java:365]: This seems like duplicity in `ConsentType` but I want to isolate possible changes in REST a provider API.

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/provider/OnboardingProvider.java [hunk 3]
```diff
@@ -32,4 +32,21 @@ public interface OnboardingProvider {
...
     void sendOtpCode(String userId, String otpCode, boolean resend) throws OnboardingProviderException;
 
+    /**
+     * Provide consent text.
+     *
+     * @param request consent text request
+     * @return consent text
+     * @throws OnboardingProviderException if there is a problem to fetch consent text
+     */
+    String fetchConsent(ConsentTextRequest request) throws OnboardingProviderException;
```
Comment (banterCZ) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/provider/OnboardingProvider.java:42]: _Context object_ approach makes keeping backward compatibility easier. We may add parameters and don't break compatibility.

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/controller/api/OnboardingController.java [hunk 4]
```diff
@@ -183,4 +187,50 @@ public Response performCleanup(@EncryptedRequestBody ObjectRequest<OnboardingCle
...
+            throw new ValidationException("Missing mandatory attributes");
+        }
+        try {
+            LocaleUtils.toLocale(request.getLanguage());
+        } catch (IllegalArgumentException e) {
+            throw new ValidationException("Invalid language: " + request.getLanguage(), e);
+        }
+    }
+
+    private static void validate(final OnboardingConsentApprovalRequest request) {
```
Comment (banterCZ) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/controller/api/OnboardingController.java:228]: Or do you prefer to replace it by _javax.validation_?

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/controller/api/OnboardingController.java (hunk 1 above)
Comment (petrdvorak) [reply enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/controller/api/OnboardingController.java:202]: This looks sufficient. 👍

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/controller/api/OnboardingController.java (hunk 4 above)
Comment (petrdvorak) [reply enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/controller/api/OnboardingController.java:228]: Typically, we would isolate the logic into a separate `OnboardingConsentApprovalRequestValidator` class.

File: enrollment-server-api-model/src/main/java/com/wultra/app/enrollmentserver/api/model/request/ConsentType.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+package com.wultra.app.enrollmentserver.api.model.request;
+
+/**
+ * Consent type.
+ *
+ * @author Lubos Racansky, lubos.racansky@wultra.com
+ */
+public enum ConsentType {
+
+    GDPR
```
Comment (petrdvorak) [inline enrollment-server-api-model/src/main/java/com/wultra/app/enrollmentserver/api/model/request/ConsentType.java:27]: I think we need to have this freeform string. Can you name consent types the customers will use? Validation can be done on some other layer.

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/OnboardingService.java (hunk 2 above)
Comment (petrdvorak) [reply enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/OnboardingService.java:365]: this method should be isolated to separate converter class.

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/OnboardingService.java [hunk 6]
```diff
@@ -316,6 +318,58 @@ public void terminateInactiveProcesses() {
...
+            throw new OnboardingProcessException("An error when fetching consent text.", e);
+        }
+    }
+
+    /**
+     * Record dis/approval of consent
+     *
+     * @param request approval request
+     */
+    public void approveConsent(final OnboardingConsentApprovalRequest request) throws OnboardingProcessException {
```
Comment (petrdvorak) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/OnboardingService.java:350]: Let's consider if we can live with void method here.

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/provider/ConsentType.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+package com.wultra.app.onboardingserver.provider;
+
+/**
+ * Consent type.
+ *
+ * @author Lubos Racansky, lubos.racansky@wultra.com
+ */
+public enum ConsentType {
+
+    GDPR
```
Comment (petrdvorak) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/provider/ConsentType.java:27]: dtto - not sure if we can live with enum here...

**Review by petrdvorak — CHANGES_REQUESTED** [review]
Please consider several changes, maily separating validator/converter logic from the service/controller classes.

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/OnboardingService.java (hunk 6 above)
Comment (banterCZ) [reply enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/OnboardingService.java:350]: 👍  You are right. From compatibility point of view, it's better to return even empty object expandable in the future.

File: enrollment-server-api-model/src/main/java/com/wultra/app/enrollmentserver/api/model/request/ConsentType.java (hunk 5 above)
Comment (romanstrobl) [reply enrollment-server-api-model/src/main/java/com/wultra/app/enrollmentserver/api/model/request/ConsentType.java]: Agree.

Comment (banterCZ) [reply enrollment-server-api-model/src/main/java/com/wultra/app/enrollmentserver/api/model/request/ConsentType.java]: Changed to String.

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/provider/ConsentType.java (hunk 7 above)
Comment (banterCZ) [reply enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/provider/ConsentType.java]: Changed to String.

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/OnboardingService.java (hunk 6 above)
Comment (romanstrobl) [reply enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/OnboardingService.java:350]: Agree with empty return type.

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/provider/OnboardingProvider.java (hunk 3 above)
Comment (romanstrobl) [reply enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/provider/OnboardingProvider.java:42]: This is a good idea, however we should stick to one approach within this interface, we use two approaches now (see methods above).

**Review by romanstrobl — APPROVED** [review]
Nice improvements. Let's finalize the migration of consent type to String and keep the interface tidy.

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/provider/ApproveConsentRequest.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.util.UUID;
+
+/**
+ * Request object for {@link OnboardingProvider#approveConsent(ApproveConsentRequest)}.
+ *
+ * @author Lubos Racansky, lubos.racansky@wultra.com
+ */
+@Builder
+@Getter
```
Comment (banterCZ) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/provider/ApproveConsentRequest.java:31]: This should be annotated by `PublicApi` when #222 merged

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/provider/OnboardingProvider.java (hunk 3 above)
Comment (banterCZ) [reply enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/provider/OnboardingProvider.java:42]: Yes, I was about make it consinstent in a separate PR.

**Pull request merged**